nntrainer/dataset/batch_queue.h

```diff
@@ -91,11 +96,184 @@ class BatchQueue {
   std::condition_variable_any q_reader_cv;
   std::condition_variable_any q_writer_cv;
 
+  std::queue<std::unique_ptr<T>> q;
+};
+
+/**
+ * @brief A view container that calls a callback on destruct
+ * @note the callback must be noexcept, and the given underlying data must
+ * outleave the lifetime of this class
+ *
+ * @tparam T underlying type
+ */
+template <typename T> class ScopedView {
+public:
   /**
-   * @todo consider using circular buffer if this is too slow
+   * @brief Construct a new Scoped View object
    *
+   * @param data_ reference of the underlying data
+   * @param on_notify_ callback to be called on exit, this is not copied but
```

Comment:
this is ok for now but later, it would be good for on_notify_  to have a `void *user_data`. Possibly add a todo here.

Reply:
on notify will never receive a parameter. It will be called when the object is being destroyed. Passing a param to constructor is possible but lambda capture is more intuitive.